python/private/py_package.bzl

```diff
@@ -31,36 +31,62 @@ def _path_inside_wheel(input_file):
     return short_path
 
 def _py_package_impl(ctx):
-    inputs = depset(
-        transitive = [dep[DefaultInfo].data_runfiles.files for dep in ctx.attr.deps] +
-                     [dep[DefaultInfo].default_runfiles.files for dep in ctx.attr.deps],
-    )
-
     # TODO: '/' is wrong on windows, but the path separator is not available in starlark.
     # Fix this once ctx.configuration has directory separator information.
     packages = [p.replace(".", "/") for p in ctx.attr.packages]
-    if not packages:
-        filtered_inputs = inputs
+    exclude = [p.replace(".", "/") for p in ctx.attr.exclude]
+    input_files = depset(transitive = [dep[PyInfo].transitive_sources for dep in ctx.attr.deps] + [dep[DefaultInfo].default_runfiles.files for dep in ctx.attr.deps])
+    imports = []
+    if not packages and not exclude:
+        filtered_inputs = input_files
+        imports = [dep[PyInfo].imports for dep in ctx.attr.deps]
     else:
         filtered_files = []
+        for dep in ctx.attr.deps:
+            add_imports = False
 
-        # TODO: flattening depset to list gives poor performance,
-        for input_file in inputs.to_list():
-            wheel_path = _path_inside_wheel(input_file)
-            for package in packages:
-                if wheel_path.startswith(package):
-                    filtered_files.append(input_file)
+            # TODO: flattening depset to list gives poor performance,
```

Comment:
What's there TODO? It looks like the underlying objects being operated on are, rightly so, depsets; how would the flattening be removed?

Reply:
This comment was here in the first place. I didn't bother removing it, and yes, there's no option.